### PR TITLE
Implement Discord message prisms and enhance DiscordLens

### PR DESCRIPTION
**Pull request by sojinmm**
# Discord Message Management Implementation

This PR introduces comprehensive Discord message handling functionality through prisms and enhances the DiscordLens implementation.

## Key Changes
- Implement SendMessagePrism for creating new messages
- Add EditMessagePrism for modifying existing messages
- Create DeleteMessagePrism for message removal
- Implement PinMessagePrism and UnpinMessagePrism for pin management
- Add BulkDeleteMessagesPrism for batch message deletion
- Add ReactToMessagePrism for message reactions
- Enhance DiscordLens to properly handle different HTTP methods and request parameters

## Technical Details
- Implement robust validation for Discord IDs, content length, and other parameters
- Add comprehensive error handling for Discord API responses
- Improve request parameter handling in DiscordLens for both GET and non-GET requests
- Add extensive test coverage for all message-related functionality

## Testing
- Add unit tests covering all message operations
- Include validation test cases for edge cases
- Test error handling scenarios
- Verify proper parameter handling for different HTTP methods

### CODE REVIEW

File: lux/config/config.exs [hunk 1]
```diff
@@ -6,6 +6,10 @@ config :lux, :open_ai_models,
   default: "gpt-4o-mini",
   smartest: "gpt-4o"
 
+# API Keys
+config :lux, :api_keys,
+  discord: System.get_env("DISCORD_BOT_TOKEN", "test_bot_token")
```
Comment (rparcus) [inline lux/config/config.exs:11]: Please check how we set config values at runtime in `runtime.exs` and using dotenvy to load values from different config files and the system as a fallback. This should move to `runtime.exs`

File: lux/lib/lux/lenses/discord_lens.ex [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  # Getting server information
+  DiscordLens.focus(%{
+    endpoint: "/guilds/:guild_id",
+    params: %{guild_id: "987654321"}
+  })
+  ```
+  """
+
+  use Lux.Lens,
+    name: "Discord API",
```
Comment (rparcus) [inline lux/lib/lux/lenses/discord_lens.ex:37]: This lens is missing a schema entirely so an LLM would have no way to know how to use it. Also, let's avoid "blanket" lenses or prisms, as those do not work well with LLM Tool Calling. Tools (Prisms, Lenses, etc) should be as minimal as possible. Do one specific thing with as few parameters as needed, so that the LLMs do not need to waste tokens in parsing the schemas.

File: lux/lib/lux/prisms/discord/messages/bulk_delete_messages_prism.ex [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+  @doc """
+  Handles the request to bulk delete messages in a Discord channel.
+  """
+  def handler(%{channel_id: channel_id, message_ids: message_ids} = input, %{agent: agent} = _ctx) do
+    Logger.info("Agent #{agent.name} bulk deleting #{length(message_ids)} messages in channel #{channel_id}")
+
+    headers = if input[:reason], do: [{"X-Audit-Log-Reason", input.reason}], else: []
+
+    case DiscordLens.focus(%{
```
Comment (rparcus) [inline lux/lib/lux/prisms/discord/messages/bulk_delete_messages_prism.ex:81]: No please let's not do this. Lenses are intended to be used by Agents, or as building blocs inside Beams. They represent a way to load "resources" (something that other framework such as MCP also try to model separately from "Actions"). If we need a discord client to use inside Prisms, let's just write one (or use a library if it is much better, but I think that for our use case we can easily build a basic discord HTTP client with Req.

File: lux/lib/lux/prisms/discord/messages/bulk_delete_messages_prism.ex [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+      {:error, "Missing required fields: channel_id, message_ids"}
+    else
+      with {:ok, _} <- validate_channel_id(input.channel_id),
+           {:ok, _} <- validate_message_ids(input.message_ids) do
+        :ok
+      end
+    end
+  end
+
+  defp validate_channel_id(channel_id) when is_binary(channel_id) do
```
Comment (rparcus) [inline lux/lib/lux/prisms/discord/messages/bulk_delete_messages_prism.ex:115]: It would be best if we could unit test all these functions. By writing a discord client to be reused, and testing it thoroughly, you make sure that the rest of the work will go smoothly.

File: lux/lib/lux/prisms/discord/messages/delete_message_prism.ex [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+  @doc """
+  Handles the request to delete a message from a Discord channel.
+  """
+  def handler(%{channel_id: channel_id, message_id: message_id} = input, %{agent: agent} = _ctx) do
+    Logger.info("Agent #{agent.name} deleting message #{message_id} from channel #{channel_id}")
+
+    headers = if input[:reason], do: [{"X-Audit-Log-Reason", input.reason}], else: []
+
+    case DiscordLens.focus(%{
```
Comment (rparcus) [inline lux/lib/lux/prisms/discord/messages/delete_message_prism.ex:77]: As mentioned above, let's not write a "fake" lens to use it as a client. Lenses are intended as tools for LLMs, not as building blocks for normal programs. It's overkill.

File: lux/lib/lux/prisms/discord/messages/edit_message_prism.ex [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            Logger.info("Successfully edited message #{message_id} in channel #{channel_id}")
+            {:ok, %{message: response}}
+          {:error, reason} ->
+            Logger.error("Failed to edit Discord message: #{inspect(reason)}")
+            handle_discord_error(reason)
+        end
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
```
Comment (rparcus) [inline lux/lib/lux/prisms/discord/messages/edit_message_prism.ex:101]: imagine something like this:

```elixir

def handler(%{channel_id: channel_id, message_id: message_id, content: content} = input, %{agent: agent} = _ctx) do
  with :ok <- validate(input),
       :ok <- Logger.info("Agent #{agent.name} editing message #{message_id} in channel #{channel_id}") do
    Lux.Integrations.Discord.Client.patch("/channels/#{channel_id}/messages/#{message_id}", %{
      content: content
    })
  end
end
```
So all you'd have to do is to implement this Lux.Integrations.Discord.Client. You could do a PR that does just that and tests and have a solid foundation to build the other parts.

File: lux/lib/lux/prisms/discord/messages/edit_message_prism.ex [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+  @doc """
+  Validates the input parameters.
+  """
+  def validate(input) do
```
Comment (rparcus) [inline lux/lib/lux/prisms/discord/messages/edit_message_prism.ex:106]: We actually have `ExJsonSchema.Validator.validate` to validate the inputs. You'd need to look at [the docs](https://github.com/jonasschmidt/ex_json_schema) to see how schema resolution works (for example used [here](https://github.com/sojinmm/lux/blob/d18fff52c8c968fc414d233fd86dff8d11348c4e/lux/lib/lux/signal_schema.ex#L85-L86) and [here](https://github.com/sojinmm/lux/blob/d18fff52c8c968fc414d233fd86dff8d11348c4e/lux/lib/lux/signal_schema.ex#L139-L140) but it should be better than writing the validation and error handling yourself.

File: lux/lib/lux/prisms/discord/messages/get_pinned_messages_prism.ex [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        }
+      },
+      required: ["messages"]
+    }
+
+  alias Lux.Lenses.DiscordLens
+  require Logger
+
+  # Discord API error codes
+  @discord_errors %{
```
Comment (rparcus) [inline lux/lib/lux/prisms/discord/messages/get_pinned_messages_prism.ex:76]: For error handling, inside a module such as `Lux.Integrations.Discord` you could have a general function that matches on all types of errors. No need to have them specified (and repeated) on each Prism/Lens I guess

File: lux/lib/lux/prisms/discord/messages/pin_message_prism.ex [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+          type: :string,
+          description: "The ID of the channel containing the message",
+          pattern: "^[0-9]{17,20}$"
+        },
+        message_id: %{
+          type: :string,
+          description: "The ID of the message to pin",
+          pattern: "^[0-9]{17,20}$"
+        },
+        reason: %{
```
Comment (rparcus) [inline lux/lib/lux/prisms/discord/messages/pin_message_prism.ex:29]: I think that since many of these arguments are optional and perhaps not really needed for most day to day operations, we might want to omit them.

File: lux/lib/lux/prisms/discord/messages/pin_message_prism.ex [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    else
+      with {:ok, _} <- validate_channel_id(input.channel_id),
+           {:ok, _} <- validate_message_id(input.message_id),
+           {:ok, _} <- validate_reason(input[:reason]) do
+        :ok
+      end
+    end
+  end
+
+  defp validate_channel_id(channel_id) when is_binary(channel_id) do
```
Comment (rparcus) [inline lux/lib/lux/prisms/discord/messages/pin_message_prism.ex:112]: All this validation code is actually very business logic heavy. We'd have to test it with unit tests... Actually, what happens if we do not validate? Would Discord return something that the LLM could understand? Perhaps we can simply scpeify the right rules in the schema and then let Discord do the validation, saving us A TON of extra code and possible bugs.

File: lux/lib/lux/prisms/discord/messages/read_channel_messages_prism.ex [hunk 11]
```diff
@@ -0,0 +1,161 @@
+defmodule Lux.Prisms.Discord.Messages.ReadChannelMessagesPrism do
```
Comment (rparcus) [inline lux/lib/lux/prisms/discord/messages/read_channel_messages_prism.ex:1]: Read channel would be the perfect thing to be implemented as a simple Lens. Why does it have to be a Prism?

File: lux/test/unit/lux/discord_lens_test.exs [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+      response = %{"data" => "test"}
+
+      with_mocks([
+        {Req, [], [
+          new: fn _opts -> {:ok, %{}} end,
+          request: fn _req, _opts ->
+            {:ok, %{status: 200, body: response}}
+          end
+        ]},
+        {Lux.Config, [], [
```
Comment (rparcus) [inline lux/test/unit/lux/discord_lens_test.exs:26]: Please check other Lenses tests where we simply use `Req.Test`. Let's not try to mock `Lux.Config`.

File: lux/test/unit/lux/prisms/messages_test.exs [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+  describe "EditMessagePrism" do
+    test "successfully edits a message" do
+      new_content = "Updated message"
+      response = %{
+        "id" => @valid_message_id,
+        "content" => new_content,
+        "edited_timestamp" => "2024-01-01T00:00:00.000Z"
+      }
+
+      with_mock DiscordLens, [
```
Comment (rparcus) [inline lux/test/unit/lux/prisms/messages_test.exs:58]: Please take a look at Req.Test and if you do the Discord.Client, you could mock that instead. I think it's better to mock Req directly so that you can actually test your code after the HTTP requests are done.

File: lux/test/unit/lux/prisms/messages_test.exs [hunk 14]
```diff
@@ -0,0 +1,238 @@
+defmodule Lux.Prisms.Discord.MessagesTest do
+  use ExUnit.Case
+  import Mock
+
+  alias Lux.Prisms.Discord.Messages.{
+    SendMessagePrism,
```
Comment (rparcus) [inline lux/test/unit/lux/prisms/messages_test.exs:6]: Please let's have one test module per Action/Lens...

**Review by rparcus — CHANGES_REQUESTED** [review]
Hello @sojinmm, thanks for your contribution! So far, I think that this one is the best of the the discord ones. 
There are still things to do, so I left different comments that should give you an idea of what works and what doesn't with this approach. 
Building a dedicate generic client to use in Prisms is probably a good idea. That could a single PR with nice tests.
After that, I'd try to build ONE prism that works with proper tests and perhaps also integrations tests.
Same thing for another PR for ONE Lenses.

One we have examples of those in a way that we are happy with, we could vibe code the rest, so that we don't need to go back and forth with a thousands of lines to review/change. 

WDYT?

**Comment by sojinmm** [comment]
> Hello @sojinmm, thanks for your contribution! So far, I think that this one is the best of the the discord ones. There are still things to do, so I left different comments that should give you an idea of what works and what doesn't with this approach. Building a dedicate generic client to use in Prisms is probably a good idea. That could a single PR with nice tests. After that, I'd try to build ONE prism that works with proper tests and perhaps also integrations tests. Same thing for another PR for ONE Lenses.
> 
> One we have examples of those in a way that we are happy with, we could vibe code the rest, so that we don't need to go back and forth with a thousands of lines to review/change.
> 
> WDYT?

That sounds like a good idea. Currently, I was thinking of just focusing on developing lenses and prisms. Using the lens via API communication led to inefficiencies in other parts of the code. However, creating a generic client seems like a much better approach. I'll start from here.

Thank you.

**Comment by rparcus** [comment]
> > Hello @sojinmm, thanks for your contribution! So far, I think that this one is the best of the the discord ones. There are still things to do, so I left different comments that should give you an idea of what works and what doesn't with this approach. Building a dedicate generic client to use in Prisms is probably a good idea. That could a single PR with nice tests. After that, I'd try to build ONE prism that works with proper tests and perhaps also integrations tests. Same thing for another PR for ONE Lenses.
> > One we have examples of those in a way that we are happy with, we could vibe code the rest, so that we don't need to go back and forth with a thousands of lines to review/change.
> > WDYT?
> 
> That sounds like a good idea. Currently, I was thinking of just focusing on developing lenses and prisms. Using the lens via API communication led to inefficiencies in other parts of the code. However, creating a generic client seems like a much better approach. I'll start from here.
> 
> Thank you.

That's great. If you have some new ideas now and want to start new work, please consider closing the PRs that are not relevant anymore. Looking forward for your next commit :) 
Thanks!
